src/safe-event-emitter-provider.ts

```diff
@@ -0,0 +1,62 @@
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import type { JsonRpcEngine, JsonRpcRequest } from 'json-rpc-engine';
+
+/**
+ * An Ethereum provider.
+ *
+ * This provider loosly follows conventions that pre-date EIP-1193. It
+ * is not compliant with any Ethereum provider standard.
+ */
+export class SafeEventEmitterProvider extends SafeEventEmitter {
+  #engine: JsonRpcEngine;
+
+  /**
+   * Construct a SafeEventEmitterProvider from a JSON-RPC engine.
+   *
+   * @param options - Options.
+   * @param options.engine - The JSON-RPC engine used to process requests.
+   */
+  constructor({ engine }: { engine: JsonRpcEngine }) {
+    super();
+    this.#engine = engine;
+
+    if (engine.on) {
+      engine.on('notification', (message: string) => {
+        this.emit('data', null, message);
+      });
+    }
+  }
+
+  /**
+   * Send a provider request asynchronously.
+   *
+   * @param req - The request to send.
+   * @param callback - A function that is called upon the success or failure of the request.
+   */
+  sendAsync = (
+    req: JsonRpcRequest<unknown>,
+    callback: (error: unknown, providerRes?: any) => void,
+  ) => {
+    this.#engine.handle(req, callback);
```

Comment:
why does the async version get a callback again?

Reply:
I'm just matching the previous API: https://github.com/MetaMask/eth-json-rpc-middleware/blob/main/src/providerFromEngine.ts#L11

We pass this provider into a bunch of third-party libraries in various different places, so making breaking changes will need to be approached carefully at a later date. I am very excited to try converting this to an EIP-1193 provider though. I have been wanting to do that forever.

Reply:
ok one other question. If the `callback` argument isn't optional should we include an error when its not thrown like there is in the legacy `send` implementation? Or is that a breaking change we don't want to risk?

Reply:
>when we do that `sendAsync` will be superceded by `request` without a callback: 

Yep! Exactly

>If the `callback` argument isn't optional should we include an error when its not thrown like there is in the legacy `send` implementation?

Good question. We could do that. I chose not to because I wanted to minimize functional changes, and because conventionally we rely on types to ensure correct parameters are passed rather than runtime checks (where possible).

I don't know why the check was added just for `send`, but I'm guessing it's because there was already an established conversion of passing a callback to `sendAsync`, whereas on `send` it was more likely it would be missing (so the improved error message was helpful).